bbpower/compsep.py

```diff
@@ -1,239 +1,364 @@
+        """
+        Hamimeche and Lewis likelihood. 
+        Taken from Cobaya written by H, L and Torrado
+        See: https://github.com/CobayaSampler/cobaya/blob/master/cobaya/likelihoods/_cmblikes_prototype/_cmblikes_prototype.py
+        """
         model_cls = self.model(params)
-        return -0.5 * np.mat(self.bbdata - model_cls) * np.mat(self.invcov) * np.mat(self.bbdata - model_cls).T
+        dx_vec = []
+        for k in range(model_cls.shape[0]):
+            C = model_cls[k] + self.bbnoise[k]
+            X = self.h_and_l(C, self.observed_cls[k], self.Cfl_sqrt[k])
+            dx = self.matrix_to_vector(X).flatten()
+            dx_vec = np.concatenate([dx_vec, dx])
+        return -0.5 * np.einsum('i, ij, j', dx_vec, self.invcov, dx_vec)
+
+    def h_and_l(self, C, Chat, Cfl_sqrt):
+        diag, U = np.linalg.eigh(C)
+        rot = U.T.dot(Chat).dot(U)
+        roots = np.sqrt(diag)
+        for i, root in enumerate(roots):
+            rot[i, :] /= root
+            rot[:, i] /= root
+        U.dot(rot.dot(U.T), rot)
+        diag, rot = np.linalg.eigh(rot)
+        diag = np.sign(diag - 1) * np.sqrt(2 * np.maximum(0, diag - np.log(diag) - 1))
+        Cfl_sqrt.dot(rot, U)
+        for i, d in enumerate(diag):
+            rot[:, i] = U[:, i] * d
+        return rot.dot(U.T)
 
     def lnprob(self, params):
+        """
+        Likelihood with priors. 
+        """
         prior = self.lnpriors(params)
         if not np.isfinite(prior):
             return -np.inf
-        lnprob = self.lnlike(params)
+        if self.use_handl:
+            lnprob = self.h_and_l_lnlike(params)
+        else:
+            lnprob = self.chi_sq_lnlike(params)
         return prior + lnprob
 
-    def emcee_LOL_sampler(self):
-        ndim, nwalkers = 8, 128
-        popt = [1., 1., 1., -3., 1.5, -0.5, -0.5, 0.5]
-        pos = [popt * (1. + 1.e-3*np.random.randn(ndim)) for i in range(nwalkers)]
-        sampler = emcee.EnsembleSampler(nwalkers, ndim, self.lnprob, args=[params])
-        sampler.run_mcmc(pos, 2**5);
+    def emcee_sampler(self):
+        """
+        Sample the model with MCMC. 
+        """
+        zmask = self.parameters.param_init == 0
+        self.parameters.param_init[zmask] += 1.e-3 * np.ones_like(zmask)
+        
+        nwalkers = self.config['nwalkers']
+        n_iters = self.config['n_iters']
+        ndim = len(self.parameters.param_init)
+        pos = [self.parameters.param_init * (1. + 1.e-3*np.random.randn(ndim)) for i in range(nwalkers)]
+
+        sampler = emcee.EnsembleSampler(nwalkers, ndim, self.lnprob)
+        sampler.run_mcmc(pos, n_iters);
         return sampler
 
+    def make_output_dir(self):
+        from datetime import datetime
+        import os, errno
+        from shutil import copyfile
+        fmt='%Y-%m-%d-%H-%M'
+        date = datetime.now().strftime(fmt)
+        output_dir = './test_bbpower_minimal/outputs/'+self.config['save_prefix']+'_'+date
```

Comment:
why are you appending `test_bbpower_minimal` to this? You could just include that in `save_prefix`, right? that way you don't have to hard-code anything

Reply:
Fixed in last commit